init.lua

```diff
@@ -60,16 +70,94 @@ local function re_run_failed_jobs(full_repo, run_id)
 function webhook_handler(req)
-    local job = req:json()
-    local run_id = job.workflow_job.run_id
-    local full_repo = job.repository.full_name
-    if is_job_failed(job) and needs_restart(run_id) then
-        fiber.create(function() re_run_failed_jobs(full_repo, run_id) end)
+    local payload = req:json()
+    if payload.workflow_job == nil then
+        log.error('Empty payload')
+        return { status = 502 }
+    end
+    if payload.workflow_job.run_attempt >= 4 then
+        log.info('Attempt >= 4, noop')
+        return { status = 204 }
+    end
+    local workflow_run_id = payload.workflow_job.run_id
+    local job_run_id = payload.workflow_job.id
+    local full_repo = payload.repository.full_name
+    local is_failed = is_job_failed(payload)
+    local is_needs_restart = false
+    if is_failed then
+        is_needs_restart = needs_restart(run_id)
+    end
+
+    describe(payload)
+
+    if payload.action == 'queued' then
+        local workflow_record = box.space.jobs:get({ workflow_run_id })
+        if workflow_record then
+            box.space.jobs:update(workflow_run_id, { { '+', 'count', 1}})
+        else
+            box.space.jobs:insert({ workflow_run_id, 1, true})
+        end
+        workflow_record = box.space.jobs:get({ workflow_run_id })
+        log.info('Job queued in workflow '..workflow_run_id..': '..tostring(workflow_record))
+        return { status = 200 }
+    elseif payload.action == 'in_progress' then
+        return { status = 200 }
+    elseif payload.action == 'completed' then
+        local workflow_record = box.space.jobs:get({ workflow_run_id })
+        if workflow_record then
+            box.space.jobs:update(workflow_run_id, { { '-', 'count', 1}})
+            if payload.workflow_job.conclusion == 'failed' then
+                box.space.jobs:update(workflow_run_id, { { '=', 'fixed', false}})
+            end
+
+            workflow_record = box.space.jobs:get({ workflow_run_id })
+            log.info('Completed job in workflow '..workflow_run_id..': '..tostring(workflow_record))
+
+            if workflow_record.count == 0 and workflow_record.fixed == false then
+                re_run_failed_jobs(full_repo, workflow_run_id)
+            end
+
+            if workflow_record.count == 0 then
+                box.space.jobs:delete(workflow_run_id)
+            end
+            return { status = 201 }
+        else
+            return { status = 404 }
+        end
+    end
+    if is_failed then
+        if is_needs_restart then
+            -- create a fiber which will repeatedly try to restart the workflow
+            fiber.create(function() re_run_failed_jobs(full_repo, run_id) end)
+            return { status = 201 }
+        else
+            -- failed, but not ready to restart
+            return { status = 202 }
+        end
+    else
+        -- no action is required
+        return { status = 200 }
     end
```

Comment:
this part was commented in f63060a924a5234ed11e19dea5023958e610e5c2 because we didn't need it anymore.

Reply:
fixed